core/txpool/legacypool/celo_list.go

```diff
@@ -0,0 +1,250 @@
+	})
+	if len(removed) == 0 {
+		return nil, nil
+	}
+	var invalids types.Transactions
+	// If the list was strict, filter anything above the lowest nonce
+	if c.list.strict {
+		lowest := uint64(math.MaxUint64)
+		for _, tx := range removed {
+			if nonce := tx.Nonce(); lowest > nonce {
+				lowest = nonce
+			}
+		}
+		invalids = c.list.txs.filter(func(tx *types.Transaction) bool { return tx.Nonce() > lowest })
+	}
+	// Reset total cost
+	c.subTotalCost(removed)
+	c.subTotalCost(invalids)
+	c.list.txs.reheap()
+	return removed, invalids
+}
+
+// Add tries to insert a new transaction into the list, returning whether the
+// transaction was accepted, and if yes, any previous transaction it replaced.
+//
+// If the new transaction is accepted into the list, the lists' cost and gas
+// thresholds are also potentially updated.
+func (c *celo_list) Add(tx *types.Transaction, priceBump uint64, l1CostFn txpool.L1CostFunc) (bool, *types.Transaction) {
+	oldNativeTotalCost := big.NewInt(0).Set(c.list.totalcost)
+	added, oldTx := c.list.Add(tx, priceBump, l1CostFn)
+	if !added {
+		return false, nil
+	}
+	if !txpool.IsFeeCurrencyTx(tx) && oldTx != nil && !txpool.IsFeeCurrencyTx(oldTx) {
+		// both the tx and the replacement are native currency, nothing to do
+		return true, oldTx
+	}
+	// undo change in native totalcost
+	c.list.totalcost.Set(oldNativeTotalCost)
+	// Recalculate
+	c.addTotalCost(tx)
+	// TODO: Add rollup cost, translated to the feecurrency of the tx
+	// TODO(hbandura): op-geth impl doesn't remove the l1cost from the removed tx
+	// is this intentional?
+	// Remove replaced tx cost
+	if oldTx != nil {
+		c.subTotalCost(types.Transactions{oldTx})
+	}
+	return added, oldTx
+}
+
+// Forward removes all transactions from the list with a nonce lower than the
+// provided threshold. Every removed transaction is returned for any post-removal
+// maintenance.
+func (c *celo_list) Forward(threshold uint64) types.Transactions {
+	txs := c.list.txs.Forward(threshold)
+	// Goes through celo_list subtotalcost to remove currency specific balances.
+	c.subTotalCost(txs)
+	return txs
+}
+
+// Cap places a hard limit on the number of items, returning all transactions
+// exceeding that limit.
+func (c *celo_list) Cap(threshold int) types.Transactions {
+	txs := c.list.txs.Cap(threshold)
+	// Goes through celo_list subtotalcost to remove currency specific balances.
+	c.subTotalCost(txs)
+	return txs
+}
+
+// Remove deletes a transaction from the maintained list, returning whether the
+// transaction was found, and also returning any transaction invalidated due to
+// the deletion (strict mode only).
+func (c *celo_list) Remove(tx *types.Transaction) (bool, types.Transactions) {
+	// Remove the transaction from the set
+	nonce := tx.Nonce()
+	if removed := c.txs.Remove(nonce); !removed {
+		return false, nil
+	}
+	// Goes through celo_list subtotalcost to remove currency specific balances.
+	c.subTotalCost([]*types.Transaction{tx})
```

Comment:
First we remove a transaction with `tx.Nonce()`, then we reduce the cost by the cost of `tx`. This assumes that the removed transaction has the same cost as `tx`. Is this guaranteed?